source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,624 @@
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Check/Process messages for incoming Cloud to Device messages.
+ *
+ * */
+static uint32_t prvAzureIoTHubClientC2DProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                void * pxPublishInfo )
+{
+    AzureIoTHubClientResult_t xResult;
+    AzureIoTHubClientCloudToDeviceMessageRequest_t xCloudToDeviceMessage = { 0 };
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pxPublishInfo;
+    az_result xCoreResult;
+    az_iot_hub_client_c2d_request xOutEmbeddedRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pcTopicName, xMQTTPublishInfo->usTopicNameLength );
+
+    /* Failed means no topic match */
+    if( az_result_failed( xCoreResult = az_iot_hub_client_c2d_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                                    xTopicSpan, &xOutEmbeddedRequest ) ) )
+    {
+        AZLogWarn( ( "Cloud to device topic parsing failed: 0x%08x", xCoreResult ) );
+        xResult = eAzureIoTHubClientTopicNoMatch;
+    }
+    else
+    {
+        /* Verify the received publish is for the feature we have subscribed to. */
+        AZLogDebug( ( "Cloud xCloudToDeviceMessage topic: %.*s  with payload : %.*s.",
+                      xMQTTPublishInfo->usTopicNameLength,
+                      xMQTTPublishInfo->pcTopicName,
+                      xMQTTPublishInfo->xPayloadLength,
+                      xMQTTPublishInfo->pvPayload ) );
+
+        if( pxContext->_internal.callbacks.xCloudToDeviceMessageCallback )
+        {
+            xCloudToDeviceMessage.pvMessagePayload = xMQTTPublishInfo->pvPayload;
+            xCloudToDeviceMessage.ulPayloadLength = ( uint32_t ) xMQTTPublishInfo->xPayloadLength;
+            xCloudToDeviceMessage.xProperties._internal.xProperties = xOutEmbeddedRequest.properties;
+
+            AZLogDebug( ( "Invoking Cloud to Device Callback." ) );
+            pxContext->_internal.callbacks.xCloudToDeviceMessageCallback( &xCloudToDeviceMessage,
+                                                                          pxContext->_internal.pvCallbackContext );
+            AZLogDebug( ( "Returned from Cloud to Device Callback." ) );
+        }
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Check/Process messages for incoming direct method messages.
+ *
+ * */
+static uint32_t prvAzureIoTHubClientDirectMethodProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                         AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                         void * pxPublishInfo )
```

Comment:
>pxPublishInfo [](start = 64, length = 13)

pv

Reply:
all the places where we have void pointer

---
In reply to: [610886237](https://github.com/Azure/azure-iot-middleware-freertos/pull/60#discussion_r610886237) [](ancestors = 610886237)